torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -374,6 +372,51 @@ def adaptiveavgpool2d_inference_rule(n: Node, module_instance):
     n.type = get_greatest_upper_bound(n.type, adaptiveavgpool2d_check(n.args[0].type, module_instance))
     return n.type
 
+def flatten_check(typ, start_dim, end_dim):
+    l = len(typ.__args__)
+
+    if 0 <= start_dim <= (l - 1) and end_dim == -1 or 0 <= end_dim <= (l - 1) and start_dim < end_dim:
```

Comment:
Can you describe what this is checking?

I suggest pre-normalizing the dimensions at the very beginning so we don't have to have extra checks for `-1`, e.g.:

```
def flatten_check(typ, start_dim, end_dim):
    l = len(typ.__args__)

    start_dim = (l - 1) if start_dim == -1 else start_dim
    end_dim = (l - 1) if end_dim == -1 else end_dim
```

Then I believe this check should just be
```
if start_dim <= end_dim:
```

Reply:
that makes sense. I was also checking if the user gave a non-existent index, for example a tensor of size 3 and they give index 10. If so I think we should raise a type error too.

Reply:
also  I used l instead of (l - 1)  because of the list comprehensions. For example [1,2,3,5,6][:5] =  [1,2,3,5,6]